File: Question 0045 - Jump Game II/src/Solution.java
import java.util.Arrays;

class Solution
{
    private int solve(int[] nums, int idx, int jump, int[] dp)
    {
        //base case
        if(idx >= nums.length -1) return 0;
        if(dp[idx] != -1) return dp[idx];

        int ans = Integer.MAX_VALUE;
        for(int i = 1; i <= jump && i+idx < nums.length; i++)
        {
            int jmp = solve(nums,idx+i, nums[idx+i], dp);
            if(jmp != Integer.MAX_VALUE) ans = Math.min(ans, 1+ jmp);
        }

        return dp[idx] = ans;
    }
    public int jump(int[] nums)
    {
        int[] dp = new int[nums.length];
        Arrays.fill(dp,-1);
        return solve(nums,0,nums[0], dp);
    }
}

File: Question 0045 - Jump Game II/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

int solve(vector<int>& nums, int idx, int val, vector<int>& dp)
{
    if(idx >= nums.size()-1) return 0;
    if(dp[idx] != -1) return dp[idx];

    int ans = INT_MAX;
    for(int i = 1; i <= val && i + idx < nums.size(); i++)
    {
        int jmp =  solve(nums, idx+i, nums[idx+i], dp);
        if(jmp != INT_MAX) ans = min(ans, 1 + jmp);
    }

    return dp[idx] = ans;
}

class Solution 
{
public:
    int jump(vector<int>& nums) 
    {
        vector<int> dp(nums.size(), -1);
        return solve(nums, 0, nums[0], dp);
    }
};